include/deal.II/hp/refinement.h

```diff
@@ -236,51 +283,35 @@ namespace hp
+     * criterion.
      *
-     * If a cell is flagged for adaptation, we will perform p-adaptation once
-     * the associated error indicators $\eta_{K}$ on cell $K$ satisfy
-     * $\eta_{K} < \eta_{K,\text{pred}}$, where the subscript $\text{pred}$
-     * denotes the predicted error. This corresponds to our assumption of
-     * smoothness being correct, else h-adaptation is supplied.
-     *
-     * For the very first adaptation step, the user needs to decide whether h-
-     * or p-adaptation is supposed to happen. An h-step will be applied with
-     * $\eta_{K,\text{pred}} = 0$, whereas $\eta_{K,\text{pred}} = \infty$
-     * ensures a p-step. The latter may be realised with
-     * `std::numeric_limits::max()`.
+     * We consider a cell for p-adaptivity if it is flagged correspondingly and
+     * its indicator successfully compares to the corresponding criterion. Other
+     * than functions p_adaptivity_from_absolute_threshold() and
+     * p_adaptivity_from_relative_threshold(), comparison functors have to be
+     * provided either via the parameters @p compare_refine and @p compare_coarsen
+     * or the corresponding template arguments.
      *
-     * Each entry of the parameter @p error_indicators and @p predicted_errors
-     * needs to correspond to an active cell.
-     *
-     * For more theoretical details see
-     * @code{.bib}
-     * @article{Melenk2001,
-     *  author    = {Melenk, Jens Markus and Wohlmuth, Barbara I.},
-     *  title     = {{On residual-based a posteriori error estimation
-     *                in hp-FEM}},
-     *  journal   = {{Advances in Computational Mathematics}},
-     *  volume    = {15},
-     *  number    = {1},
-     *  pages     = {311--331},
-     *  publisher = {Springer US},
-     *  year      = {2001},
-     *  doi       = {10.1023/A:1014268310921}
-     * }
-     * @endcode
+     * Each entry of the parameters @p indicators and @p criteria needs to
+     * correspond to an active cell.
      *
      * @note Preceeding calls of Triangulation::prepare_for_coarsening_and_refinement()
      *   may change refine and coarsen flags, which will ultimately change the
      *   results of this function.
      */
-    template <int dim, typename Number, int spacedim>
+    template <int dim,
+              typename Number,
+              int spacedim,
+              typename ComparatorRefine,
+              typename ComparatorCoarsen>
     void
-    p_adaptivity_from_prediction(
+    p_adaptivity_from_criteria(
       const hp::DoFHandler<dim, spacedim> &dof_handler,
-      const Vector<Number> &               error_indicators,
-      const Vector<Number> &               predicted_errors);
-
+      const Vector<Number> &               indicators,
+      const Vector<Number> &               criteria,
+      const ComparatorRefine &             compare_refine = ComparatorRefine(),
+      const ComparatorCoarsen &compare_coarsen = ComparatorCoarsen());
```

Comment:
For this particular function, I want the user to provide the comparators himself as template arguments to avoid ambiguity. This was intentional.

But I agree that this is not easily readable (also in terms of the documentation). If we move to std::function instead of templates, those will not be default parameters any longer.

Reply:
You can still provide defaults if you move to `std::function` objects.

Reply:
Sure. We provide default comparators for the other functions.

But for this particular one, I would prefer that the user must choose comparators, depending on how he has to deal with the reference he provided. There is just no generalization or default, and he needs to supply comparators on his own.